wcfsetup/install/files/lib/util/UnfurlUrlUtil.class.php

```diff
@@ -0,0 +1,254 @@
+<?php
+
+namespace wcf\util;
+
+use GuzzleHttp\Exception\BadResponseException;
+use GuzzleHttp\Exception\TooManyRedirectsException;
+use GuzzleHttp\Exception\TransferException;
+use GuzzleHttp\Psr7\Request;
+use GuzzleHttp\RequestOptions;
+use wcf\system\io\HttpFactory;
+
+/**
+ * Helper class to unfurl specific urls.
+ *
+ * @author      Joshua Ruesweg
+ * @copyright   2001-2021 WoltLab GmbH
+ * @license     GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package     WoltLabSuite\Core\Util
+ * @since       5.4
+ */
+final class UnfurlUrlUtil
+{
+    /**
+     * 10 Mebibyte
+     */
+    private const MAX_SIZE = (10 * (1 << 20));
+
+    /**
+     * 3 Mebibyte
+     */
+    private const MAX_IMAGE_SIZE = (3 * (1 << 20));
+
+    /**
+     * @var string
+     */
+    private $url;
+    
+    /**
+     * @var string
+     */
+    private $body;
+    
+    /**
+     * @var \DOMDocument
+     */
+    private $domDocument;
+    
+    public function __construct(string $url)
+    {
+        if (!Url::is($url)) {
+            throw new \InvalidArgumentException('Given URL "' . $url . '" is not a valid URL.');
+        }
+        
+        $this->url = $url;
+        
+        $this->fetchUrl();
+    }
+    
+    /**
+     * Fetches the body of the given url and converts the body to utf-8.
+     */
+    private function fetchUrl(): void
+    {
+        try {
+            $client = HttpFactory::makeClient([
+                RequestOptions::TIMEOUT => 10,
+                RequestOptions::STREAM => true,
+            ]);
+            $request = new Request('GET', $this->url, [
+                'range' => \sprintf('bytes=%d-%d', 0, self::MAX_SIZE - 1),
```

Comment:
Note to self: I'd like to check whether we should use a custom `user-agent` here.